### PR TITLE
grammars: upgrade to llguidance 0.7.10

### PR DESCRIPTION
**Pull request by mmoskal**
This updates [llguidance](https://github.com/guidance-ai/llguidance) from 0.6.15 to 0.7.10, notable changes:
- `%regex { ... }` syntax in Lark allowing for [substring](https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md#substring) matching (citations, code edits)
- `"x-guidance": {...}` key in JSON schema to [modify options](https://github.com/guidance-ai/llguidance/blob/main/docs/json_schema.md#whitespace-handling)
- `suffix=...` support in Lark syntax for [lazy matching](https://github.com/guidance-ai/llguidance/blob/main/docs/syntax.md#lazy-lexemes) (eg., some tool calling)
- gbnf conversion fixes
- support for `rollback()` operation on grammar for spec-decoding; not used in llama.cpp for now
- a bunch of fixes and performance improvements

The PR also switches from `llg_constraint_*()` C APIs to `llg_matcher_*()`. The new APIs are simpler.

I've also added test using sampler chain.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

File: tests/test-grammar-llguidance.cpp [hunk 1]
```diff
@@ -1086,6 +1086,65 @@ static void test_json_schema() {
         });
 }
 
+static void one_hot(llama_token_data_array & tok_arr, llama_token selected) {
```
Comment (ggerganov) [inline tests/test-grammar-llguidance.cpp:1089]: nit: is this supposed to be `one_shot`?

Comment (mmoskal) [reply tests/test-grammar-llguidance.cpp:1089]: I meant https://en.wikipedia.org/wiki/One-hot though I guess this is really logits that will eventually result in 1-hot probability
